packages/flutter/lib/src/material/icon_button.dart

```diff
@@ -778,10 +789,41 @@ class _IconButtonM3 extends ButtonStyleButton {
     return _IconButtonDefaultsM3(context);
   }
 
-  /// Returns null because [IconButton] doesn't have its component theme.
+  /// Returns the [IconButtonThemeData.style] of the closest [IconButtonTheme] ancestor.
+  /// The color and icon size can also be configured by the [IconTheme] if the same property
+  /// has a null value in [IconButtonTheme]. However, if any of the properties exist
+  /// in both [IconButtonTheme] and [IconTheme], [IconTheme] will be overridden.
   @override
   ButtonStyle? themeStyleOf(BuildContext context) {
```

Comment:
I think this function reduces to:
```dart
final IconThemeData iconTheme = IconTheme.of(context).iconTheme;
final bool isDark = Theme.of(context).brightness == Brightness.dark;
final bool isDefaultColor = identical(iconTheme.color, isDark ? kDefaultIconLightColor : kDefaultIconDarkColor);
final bool isDefaultSize = iconTheme.size == const IconThemeData.fallback().size();
final ButtonStyle iconThemeStyle = IconButton.styleFrom(
  foregroundColor: isDefaultColor ? null : iconTheme.color;
  iconSize: isDefaultSize: isDefaultSize ? null : iconTheme.size,
);

return IconButtonTheme.of(context).style?.merge(iconThemeStyle) ?? iconThemeStyle;
```

I haven't tested it (or even verified that it compiles) so I could be wrong :-). 

One aspect of this that we should confirm is that IconButton never applied the IconTheme's opacity to the icon's color. That certainly holds true here.

Reply:
Yes, it's much cleaner. Thanks for the suggestion. And the icon opacity is handled in Icon.dart, so if users want to change opacity of the icon, they can use `IconTheme.merge()`, or give a `foregroundColor.withOpacity()` in `IconButtonTheme`.